bin/guard

```diff
@@ -2,31 +2,114 @@
 
 require "pathname"
 
-default = Pathname(ENV.fetch("BUNDLE_GEMFILE", "Gemfile"))
-relative_to_binstub = Pathname(__FILE__).realpath + "../../Gemfile"
-gemfile = [default, relative_to_binstub].detect(&:exist?)
-
-if gemfile && !ENV["RUBYGEMS_GEMDEPS"]
-  ENV["BUNDLE_GEMFILE"] = gemfile.to_s
-  require "rubygems"
-  require "bundler/setup"
-else
-  require "rubygems"
-end
+class GuardReloader
+  class Config
+    def using_rubygems?
+      ENV["RUBYGEMS_GEMDEPS"]
```

Comment:
Yes. (Portions of RubyGems used to be compiled-into Ruby - gems prelude code, and --disable-gems used to disable it - and to allow using newer/alternative versions of rubygems).

My point is: I'd expect requiring rspec to fail if it isn't in the Gemfile (which is what happens when using Bundler). I'd expect a Gemfile to "limit" the gems available, no just "limit versions" of "some gems" - so a Gemfile works very differently for Bundler vs RubyGems. 

I do believe requiring RubyGems is "correct" if the RUBYGEM_GEMDEPS variable is set - even if it isn't practically necessary (unless `--disable-gems` is given, or there's another use case I haven't thought of). It doesn't hurt to require rubygems, while using guard in a pure non-rubygems context means really complex managing of ruby's `-I` parameters (to set the LOAD_PATH).

I think in Guard 3.x it may be ok to just leave the binstub section - but I'm worried users will hit edge cases (e.g. minitest users, who often rely on require paths alone).

Also, some users may have ruby 1.8 or 1.9 still installed by default on their systems (devops-related work mostly), even though they may be using user-installed 2.x Rubies (managed with rbenv/rvm) - there are cases when using setuid/root (not uncommon in docker) can create hard to track edge cases, especially with tools such as Chef or Puppet. (AFAIR, Ruby 2.0 is the default only since Yosemite - so people using anything else may still encounter pre-2.x problems).

Not to mention things like cron, where relying on RUBYGEMS_GEMDEPS seems like a brittle option (given RUBYGEMS_GEMDEPS is in the login shell startup scripts). Bundler is stricter and thus more reliable for robust configurations and scripts.

Problems with RubyGems/Bundler are a time sink every time - and the number of possible configurations out there are insane. I prefer to be on the safe side by requiring bundler and rubygems in this bin script - if you have a scenario where requiring these causes a failure (or eats up too much startup time), let me know. So I prefer "safe/defensive" code (in Guard 2.x at least), rather than having "nicely refactored code" here - unless it actually breaks something (which means it's a bug and should be fixed asap).

Another thing is that rubygems often silently avoids failures - e.g. if RUBYGEMS_GEMDEPS is set to a non-existing file, you won't get an error. It's maybe more "user friendly", but for maintainers like me, it's a nightmare of unproductive hours wasted on something that should've immediately resulted in a stack trace dump.

Also, both RubyGems and Bundler are in active development - every now and then bugs are resolved and sometimes new bugs are introduced. Being on the "safe side" and avoiding edge cases is a good idea (e.g. preventing the --disable-gems from triggering such edge cases).

I'm more worried the rest of guard (or it's plugins) are not supporting the use case you need (like the missing Pathname requires you had in your PR).

Thanks for taking the time to review these changes - I'll sleep much better being more aware of the differences between Bundler, with RubyGems and without. It would probably be best to get the test suite working without Bundler - though I'm not sure that's time well spent (dependencies between files are a mess due to deprecations).